scripts/uploader.rb

```diff
@@ -35,46 +34,174 @@ def run
+
+        upload_headers = headers.merge("ETag" => options[:source_hash])
+        upload_headers["X-Delete-At"] = options[:expires].to_i.to_s if options[:expires]
+
+        destination_url = url(destination)
 
         RestClient.put(
           destination_url,
-          File.read(source_name),
+          File.read(source),
           upload_headers
         )
 
-        puts "Uploading #{appliance} as #{destination_name}...complete: #{destination_url}"
+        puts "Uploading #{appliance} to Rackspace as #{destination}...complete: #{destination_url}"
+      end
 
-        next unless master?(appliance)
+      def copy(source, destination)
+        appliance = File.basename(source)
+        puts "Copying   #{appliance} to #{destination} on Rackspace..."
 
-        if nightly?
-          devel = devel_filename(destination_name)
-          puts "Copying   #{appliance} to #{devel}..."
+        RestClient::Request.execute(
+          :method  => :copy,
+          :url     => url(source),
+          :headers => token_headers.merge("Destination" => "/#{container}/#{destination}")
+        )
 
-          RestClient::Request.execute(
-            :method  => :copy,
-            :url     => destination_url,
-            :headers => token_headers.merge("Destination" => "/#{container}/#{devel}")
-          )
+        puts "Copying   #{appliance} to #{destination} on Rackspace...complete"
+      end
+
+      private
+
+      def url(file = nil)
+        parts = [public_url, container]
+        parts << URI.encode(file) if file
+        parts.join("/")
+      end
+
+      def public_url
+        @public_url ||= login["access"]["serviceCatalog"].detect { |i| i["name"] == "cloudFiles" }["endpoints"].detect { |i| i["region"] == region.to_s.upcase }["publicURL"]
+      end
 
-          puts "Copying   #{appliance} to #{devel}...complete"
+      def headers
+        @headers ||= token_headers.merge("X-Detect-Content-Type" => "True").freeze
+      end
+
+      def token_headers
+        @token_headers ||= {"X-Auth-Token" => login["access"]["token"]["id"]}.freeze
+      end
+    end
+
+    class DigitalOcean
+      attr_reader :access_key, :secret_key, :endpoint, :bucket
+
+      def initialize(config)
+        @bucket     = "releases-manageiq-org"
+        @access_key = config[:access_key]
+        @secret_key = config[:secret_key]
+        @endpoint   = config[:endpoint]
+      end
+
+      def upload(source, destination, options)
+        appliance = File.basename(source)
+        puts "Uploading #{appliance} to DigitalOcean as #{destination}..."
+
+        put_options = {
+          :acl    => "public-read",
+          :bucket => bucket,
+          :key    => destination,
+        }.tap { |h| h[:expires] = options[:expires] if options[:expires] }
+
+        response = nil
+        File.open(source, 'rb') do |content|
+          response = client.put_object(put_options.merge(:body => content))
```

Comment:
Think you can do

```ruby
response = File.open(...) do |content|
  client.put_pbject(...)
end
```

Reply:
Yep, not what I thought it would return, but works for me :smile: